lib/utils/Integers.ml

```diff
@@ -0,0 +1,207 @@
+  val of_int : int -> t option
+  val encode : t -> int32
+end
+
+module Int24 = struct
+  type t = int32
+
+  let zero = 0l
+  let of_int i = if is_int24 i then Some (Int32.of_int i) else None
+  let encode = Fun.id
+end
+
+(** Module type of an offset in the program *)
+module type Offset = sig
+  type t
+
+  val of_int : int -> t option
+  (** Build an offset from an integer *)
+
+  val fit_in_int24 : t -> bool
+  val to_int24 : t -> Int24.t option
+  val pp : Format.formatter -> t -> unit
+end
+
+module Offset = struct
+  type t = int
+
+  let min_offset = first_address - last_address
+  let max_offset = last_address - first_address
+  let is_valid i = min_offset <= i && i <= max_offset
+  let of_int i = if is_valid i then Some i else None
+  let fit_in_int24 t = is_int24 t
+  let to_int24 t = Int24.of_int t
+  let pp ppf t = Format.fprintf ppf "%d" t
+end
+
+(** Module type of an immediate value in the program, ie. a 32bit constant. *)
+module type Immediate = sig
+  type t
+
+  val of_int : int -> t option
+  (** Build an immediate from an integer *)
+
+  val to_uint16 : t -> UInt16.res
+  (** Convert it to a UInt16 *)
+
+  val to_word : t -> Word.t
+  (** Convert it to a word *)
+end
+
+module Immediate = struct
+  type t = int
+
+  let of_int i = if is_int32 i || is_uint32 i then Some i else None
+  let to_uint16 i = UInt16.of_int32 (Int32.of_int i) (* TODO : Check This ! *)
+  let to_word i = Int32.of_int i
+end
+
+(** A 32bit address in the program *)
+module type Address = sig
+  type t
+
+  val first : t
+  val last : t
+  val of_int : int -> t option
+  val next : t -> t option
+  val to_uint16 : t -> UInt16.res
+  val fit_in_uint16 : t -> int -> bool
+  val diff : t -> t -> Offset.t
+  val with_offset : t -> Offset.t -> t
+  val well_defined : t -> Offset.t -> bool
+  val pp : Format.formatter -> t -> unit
+  val is_after : t -> t -> bool
+
+  module Set : Set.S with type elt = t
+
+  type set = Set.t
+
+  module Map : Map.S with type key = t
+
+  type 'a map = 'a Map.t
+end
+
+module Address = struct
+  type t = int
+
+  let first = first_address
+  let last = last_address
+  let of_int i = if first_address <= i && i <= last then Some i else None
+  let next t = of_int (t + 1)
+  let to_uint16 i = UInt16.of_int32 (Int32.of_int i)
+
+  let fit_in_uint16 t i =
+    match of_int (t + i) with Some i -> is_uint16 i | None -> false
+
+  let diff a1 a2 = a1 - a2
+
+  (** Note: The normalization of the address is implementations defined.
+       Here we take the address modulo (first_address - last_address).
+       But other implementations are possibles ! *)
+  let rec normalize_addr addr =
+    if addr < first_address then
+      (* The address too small, we wrap-around to have a (very) large address *)
+      normalize_addr (addr + last_address)
+    else if addr > last_address then
+      (* The address too large, we wrap-around to have a (very) small address *)
+      normalize_addr (addr - last_address)
+    else addr
```

Comment:
Pas vraiment compris. Si il l'adresse est trop petite (dans quel sens?) ou trop grande, il devrait pas y avoir une erreur ?

Reply:
- Je serai pour une erreur lorsque le résultat de l'offset d'une adresse n'est pas dans la plage, mais j'ai pensé que c'était le genre de truc que tu n'aimerais pas ! Ça veut par contre dire qu'il n'y a pas de `jmp -1` comme première instruction qui fasse sauter le programme à la dernière adresse des programmes possibles.

- L'adresse est trop grande lorsqu'elle dépasse la plus grande adresse possible, dans ce cas, je la ramène pas à pas dans l'intervalle en soustrayant la **plage d'adresse _(Code modifier !)_**. C'est ce qui permet que :
`Adresse(2^32 - 1) + Offset(1) = Adresse (0)`


- L'adresse est trop petite lorsqu'elle est avant la plus petite adresse possible, dans ce cas, je la ramène pas à pas dans l'intervalle en ajoutant **plage d'adresse _(Code modifier !)_**. C'est ce qui permet que :
`Adresse(0) + Offset(-1) = Adresse (2^32 - 1)`

Reply:
Compris !
Finalement, je suis pas contre une erreur dans ce cas. L'utilisateur n'as pas raison de faire ça. Ou alors, un avertissement et on fait rien : le processeur CPUlm gère théoriquement déjà ce cas naturellement.

Reply:
Fixed in bca7d9cce54d4c6c746fa336784868d00579b141.